File: py/application/games/sequencium.py
import copy
from .game import Game

"""
  Example game state:
  {
    grid: [
      [Square, None, None, None, None, None],
      [None, Square, None, None, None, None],
      [None, None, Square, None, None, None],
      [None, None, None, Square, None, None],
      [None, None, None, None, Square, None],
      [None, None, None, None, None, Square],
    ],
    lastMove: {
      row: rowId,
      col: colId,
    },
    playerTurn: 1, // None, 1 or 2
    gameEnd: None,  # {win: 1}, {win: 2}, {draw: True}
  }
 
  typedef Square = {
    playerIndex: 0, // or 1
    value: 1, // 2, 3, 4, ...
    from: 'SW', // N, NE, E, SE, S, SW, W, NW, None
  }

  Example action:
  {
    playerIndex: 0,
    rowFrom: 1,
    colFrom: 1,
    rowTo: 2,
    colTo: 2,
  }
"""


_directionHelper = [ [None, 'W', 'E'], ['N', 'NW', 'NE'], ['S', 'SW', 'SE'], ]


class Sequencium(Game):

  def gameEndCondition(self, gameState):
    grid = gameState['grid']
    scores = [0, 0]
    for r in range(len(grid)):
      for c in range(len(grid[0])):
        sq = grid[r][c]
        if sq is None:
          return None
        i = sq['playerIndex']
        scores[i] = max(scores[i], sq['value'])
    result = {'scores': scores}
    if scores[0] == scores[1]:
      result['draw'] = True
    else:
      result['win'] = 0 if scores[0] > scores[1] else 1
    return result

  def getDirection(self, rowFrom, colFrom, rowTo, colTo):
    return _directionHelper[rowTo - rowFrom][colTo - colFrom]

  def playerHasAvailableMove(self, playerIndex, grid):
    for r in range(len(grid)):
      for c in range(len(grid[0])):
        if grid[r][c]:
          continue
        for dr in range(-1, 2):
          for dc in range(-1, 2):
            if (r + dr < 0 or r + dr >= len(grid) or
                c + dc < 0 or c + dc >= len(grid[0])):
              continue
            if (grid[r + dr][c + dc] and
                grid[r + dr][c + dc]['playerIndex'] == playerIndex):
              return True
    return False

  def nextPlayerTurn(self, gameState, gameSettings):
    playerIndex = gameState['activePlayerIndex']
    if playerIndex is None:
      Game.nextPlayerTurn(self, gameState)
      return
    grid = gameState['grid']
    if not self.playerHasAvailableMove(1 - playerIndex, grid):
      return
    if not self.playerHasAvailableMove(playerIndex, grid):
      Game.nextPlayerTurn(self, gameState)
      return
    if not gameSettings or not gameSettings['doubleMoves']:
      Game.nextPlayerTurn(self, gameState)
      return
    numOccupiedSquares = 0
    for r in range(len(grid)):
      for c in range(len(grid[0])):
        numOccupiedSquares += 1 if grid[r][c] else 0
    if numOccupiedSquares % 2:
      Game.nextPlayerTurn(self, gameState)

  def getInitialGameState(self, gameSettings=None):
    return {
      'grid': [
        [_sq(0, 1, None), None, None, None, None, None],
        [None, _sq(0, 2, 'NW'), None, None, None, None],
        [None, None, _sq(0, 3, 'NW'), None, None, None],
        [None, None, None, _sq(1, 3, 'SE'), None, None],
        [None, None, None, None, _sq(1, 2, 'SE'), None],
        [None, None, None, None, None, _sq(1, 1, None)],
      ],
      'lastMove': {}, # row: 3, col: 3,
      'activePlayerIndex': None,
    }
 
  def getSettingsConfig(self):
    return [
      {
        'canonicalName': 'doubleMoves',
        'displayName': 'Players move twice',
        'description': (
            'Starting with the second player’s first turn, each player '
            'moves twice per turn.'),
        'values': [True, False],
        'defaultValue': False,
      },
    ]

  def action(self, gameState, action, gamePhase=None, gameSettings=None):
    grid = gameState['grid']
    playerIndex = action['playerIndex']
    rowFrom = action['rowFrom']
    colFrom = action['colFrom']
    rowTo = action['rowTo']
    colTo = action['colTo']
    if gameState['activePlayerIndex'] is None:
      return gameState
    if (playerIndex != gameState['activePlayerIndex'] or
        rowFrom - rowTo < -1 or rowFrom - rowTo > 1 or
        colFrom - colTo < -1 or colFrom - colTo > 1 or
        grid[rowFrom][colFrom] is None or
        grid[rowFrom][colFrom]['playerIndex'] != playerIndex or
        grid[rowTo][colTo]):
      return gameState
    direction = self.getDirection(rowFrom, colFrom, rowTo, colTo)
    value = grid[rowFrom][colFrom]['value'] + 1
    newGameState = copy.deepcopy(gameState)
    newGameState['grid'][rowTo][colTo] = {
        'playerIndex': playerIndex, 'value': value, 'from': direction}
    newGameState['lastMove'] = {'row': rowTo, 'col': colTo}
    self.checkGameEndCondition(newGameState)
    self.nextPlayerTurn(newGameState, gameSettings)
    return newGameState



def _sq(playerIndex, value, direction):
	return {'playerIndex': playerIndex, 'value': value, 'from': direction}

File: node/games/Sequencium.ts
import {Game} from '../stuff/Game'
import {SettingsConfig} from '../stuff/SettingsConfig'

enum Dir {
  N, NE, E, SE, S, SW, W, NW
}


/*
Example Square:
{
  playerIndex: 0, // or 1, 2, 3
  value: 1, // 2, 3, 4, ...
  from: 'SW', // N, NE, E, SE, S, SW, W, NW, None
}
 */


interface SequenciumSquare {
  playerIndex: number
  value: number
  from: Dir | null
}

/*
  Example game state:
  {
    grid: [
      [Square, None, None, None, None, None],
      [None, Square, None, None, None, None],
      [None, None, Square, None, None, None],
      [None, None, None, Square, None, None],
      [None, None, None, None, Square, None],
      [None, None, None, None, None, Square],
    ],
    turnSequence: [],  # keeps track of playerIndex for each turn
    lastMove: {
      row: rowId,
      col: colId,
    },
    gameEnd: None,  # {win: 1}, {win: 2}, {draw: True}
  }
*/

interface SequenciumGameState {
  grid: (SequenciumSquare | null)[][]
  turnSequence: number[]
  lastMove: { row: number, col: number } | null
  gameEnd: { win: number } | { draw: true } | null
  activePlayerIndex: number | null
}

/*
Example action:
{
  playerIndex: 0,
  rowFrom: 1,
  colFrom: 1,
  rowTo: 2,
  colTo: 2,
 */

interface SequenciumAction {
  playerIndex: number
  rowFrom: number
  colFrom: number
  rowTo: number
  colTo: number
}

interface SequenciumSettings {
  gridSize: number
  doubleMoves: boolean
  "players:playerCount": number
  players: any[]
}

export class Sequencium extends Game<SequenciumGameState, SequenciumAction, SequenciumSettings> {
  canonicalName = "sequencium"

  // Ooh, look at us. We get to index arrays using negative numbers. We're so cool.
  directionHelper = [
    [Dir.NW, Dir.N, Dir.NE],
    [Dir.W, null, Dir.E],
    [Dir.SW, Dir.S, Dir.SE],
  ]

  private getDirection(rowFrom: number, colFrom: number, rowTo: number, colTo: number): Dir | null {
    return this.directionHelper[rowFrom - rowTo + 1][colFrom - colTo + 1]
  }

  action(action: SequenciumAction): boolean {
    const {playerIndex, rowFrom, colFrom, rowTo, colTo} = action
    const grid = this.gameState.grid
    if (this.gameState.activePlayerIndex === null ||
      playerIndex !== this.gameState.activePlayerIndex ||
      rowFrom - rowTo < -1 || rowFrom - rowTo > 1 ||
      colFrom - colTo < -1 || colFrom - colTo > 1 ||
      grid[rowFrom][colFrom] === null ||
      grid[rowFrom][colFrom]!.playerIndex !== playerIndex ||
      grid[rowTo][colTo] !== null) {
      return false
    }
    const direction = this.getDirection(rowFrom, colFrom, rowTo, colTo)
    const value = grid[rowFrom][colFrom]!.value + 1
    const newGameState = JSON.parse(JSON.stringify(this.gameState))
    newGameState.grid[rowTo][colTo] = {playerIndex, value, from: direction}
    newGameState.lastMove = {row: rowTo, col: colTo}
    newGameState.turnSequence.push(playerIndex)
    this.gameState = newGameState
    this.checkGameEndCondition()
    this.nextPlayerTurn()
    return true
  }

  override nextPlayerTurn(): void {
    const playerIndex = this.gameState.activePlayerIndex
    const numPlayers = this.gameSettings.players.length
    const grid = this.gameState.grid
    const turnSequence = this.gameState.turnSequence

    if (playerIndex === null) {
      super.nextPlayerTurn()
      return
    }
    if (turnSequence.length <= 1) {
      super.nextPlayerTurn()
      return
    }

    let isGridFull = true
    for (let row of grid) {
      for (let cell of row) {
        if (!cell) {
          isGridFull = false
        }
      }
    }
    if (isGridFull) return

    if (!(this.gameSettings.doubleMoves && new Set(turnSequence.slice(-2)).size > 1)) {
      super.nextPlayerTurn()
    }

    while (!this.playerHasAvailableMove()) {
      super.nextPlayerTurn()
    }
  }

  gameEndCondition(): any {
    const grid = this.gameState.grid
    const scores = this.gameSettings.players.map(() => 0)
    for (let row of grid) {
      for (let cell of row) {
        if (!cell) {
          return null
        }
        scores[cell.playerIndex] = Math.max(scores[cell.playerIndex], cell.value)
      }
    }
    const result: any = {scores}
    const winners = scores
      .map((score, i) => [i, score])
      .filter(([i, score]) => score === Math.max(...scores))

    if (winners.length > 1)
      result.draw = true
    else
      result.win = winners[0][0]

    return result
  }

  private playerHasAvailableMove(): boolean {
    const playerIndex = this.gameState.activePlayerIndex
    const grid = this.gameState.grid
    for (let r = 0; r < grid.length; r++) {
      for (let c = 0; c < grid[0].length; c++) {
        if (grid[r][c]) continue
        for (let dr = -1; dr <= 1; dr++) {
          for (let dc = -1; dc <= 1; dc++) {
            if (r + dr < 0 || r + dr >= grid.length || c + dc < 0 || c + dc >= grid[0].length) continue
            if (grid[r + dr][c + dc]?.playerIndex === playerIndex) return true
          }
        }
      }
    }
    return false
  }

  initialGameState(settings: SequenciumSettings): SequenciumGameState {
    return {
      grid: this.makeGrid(settings.gridSize, settings.players.length),
      turnSequence: [],
      lastMove: {row: -1, col: -1},
      gameEnd: null,
      activePlayerIndex: null,
    }
  }

  private makeGrid(gridSize: number, playerCount: number): (SequenciumSquare | null)[][] {
    let grid: (SequenciumSquare | null)[][] = []
    for (let i = 0; i < gridSize; i++) {
      grid.push([])
      for (let j = 0; j < gridSize; j++) {
        grid[i].push(null)
      }
    }
    let playerCorners: Dir[] = []
    if (playerCount === 2) {
      playerCorners = [Dir.NW, Dir.SE]
    } else if (playerCount === 4) {
      playerCorners = [Dir.NW, Dir.NE, Dir.SE, Dir.SW]
    }
    for (let i = 0; i < playerCorners.length; i++) {
      let row = playerCorners[i] === Dir.NW || playerCorners[i] === Dir.NE ? 0 : gridSize - 1
      let col = playerCorners[i] === Dir.NW || playerCorners[i] === Dir.SW ? 0 : gridSize - 1
      for (let val = 1; val <= Math.floor(gridSize / 2); val++) {
        grid[row][col] = this.sq(i, val, val > 1 ? playerCorners[i] : null)
        row += playerCorners[i] === Dir.NW || playerCorners[i] === Dir.NE ? 1 : -1
        col += playerCorners[i] === Dir.NW || playerCorners[i] === Dir.SW ? 1 : -1
      }
    }
    return grid
  }

  private sq(playerIndex: number, value: number, direction: Dir | null): SequenciumSquare {
    return {playerIndex, value, from: direction}
  }

  override settingsConfig(): SettingsConfig[] {
    return [
      {
        'canonicalName': 'doubleMoves',
        'displayName': 'Players move twice',
        'description': 'Starting with the second player’s first turn, each player moves twice per turn.',
        'values': [true, false],
        'defaultValue': false,
      },
      {
        'canonicalName': 'gridSize',
        'displayName': 'Grid size',
        'values': [6, 8, 10],
        'defaultValue': 6,
      },
      {
        'canonicalName': 'players:playerCount',
        'displayName': 'Player count',
        'values': [2, 4],
        'defaultValue': 2,
      }
    ]
  }
}

